File: 0115-distinct-subsequences/0115-distinct-subsequences.js
/**
 * @param {string} s
 * @param {string} t
 * @return {number}
 */
var numDistinct = function (s, t) {
    // let map = {}
    // var again = function (i, j) {
    //     if (j == t.length) return 1
    //     if (i == s.length) return 0
    //     let key = `${i},${j}`
    //     if (key in map) {
    //         return map[key]
    //     }
    //     if (s[i] == t[j]) {
    //         map[key] = again(i + 1, j + 1) + again(i + 1, j)
    //     }
    //     else {
    //         map[key] = again(i + 1, j)
    //     }
    //     return map[key]
    // }
    // return again(0, 0)

    let dp = Array(t.length + 1).fill(0)
    dp[0] = 1
    for (let i = 1; i <= s.length; i++) {
        for (let j = t.length; j >= 1; j--) {
            if (s[i - 1] == t[j - 1]) {
                dp[j] += dp[j - 1]
            }
        }
    }
    return dp[t.length]
};

File: 0115-distinct-subsequences/0115-distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        cache = {}
        def again(i,j):
            if j == len(t):
                return 1
            if i == len(s):
                return 0
            if (i,j) in cache:
                return cache[(i,j)]
            if s[i] == t[j]:
                cache[(i,j)] = again(i+1,j+1) + again(i+1,j)
            else:
                cache[(i,j)] = again(i+1,j)
            return cache[(i,j)]
        return again(0,0)